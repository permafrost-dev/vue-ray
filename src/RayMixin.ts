import { VueRay } from '@/VueRay';
import { determineComponentNameDuringEvent } from '@/lib/helpers';
import { watch } from 'vue';

const conditionallyDisplayEvent = (eventName: string, options: Record<string, unknown>, rayInstance: any = null) => {
    if (VueRay.shouldDisplayEvent(eventName)) {
        rayInstance = rayInstance ?? RayMixin.methods.$ray;

        // don't display 'unknown' components
        if (!(options?.__file ?? false)) {
            return;
        }

        const componentName: string = determineComponentNameDuringEvent(options);

        let r = rayInstance;

        if (typeof rayInstance === 'function') {
            r = rayInstance();
        }

        r.table([
            `component ${eventName}: <code>${componentName}</code>`,
            `filename: <code>&lt;project root&gt;/${options?.__file ?? 'unknown.js'}</code>`,
        ]);
    }
};

export const RayMixin = {
    beforeMount(component: any) {
        conditionallyDisplayEvent('before-mount', component);
    },

    beforeUnmount(component: any) {
        conditionallyDisplayEvent('before-unmount', component);
    },

    created(component: any) {
        conditionallyDisplayEvent('created', component);
    },

    mounted(component: any) {
        conditionallyDisplayEvent('mounted', component);
    },

    unmounted(component: any) {
        conditionallyDisplayEvent('unmounted', component);
    },

    updated(component: any) {
        conditionallyDisplayEvent('updated', component);
    },

    methods: {
        $ray(...args: any[]) {
            const ray = VueRay.create();
            ray.component = this;
            ray.watch = watch;

            return ray.send(...args);
        },
    },
};
