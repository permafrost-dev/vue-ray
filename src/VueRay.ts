import { determineComponentNameDuringEvent } from '@/lib/helpers';
import { RayMixin } from '@/RayMixin';
import { format as prettyPrint } from '@permafrost-dev/pretty-format';
import { Ray } from 'node-ray/dist/web';
import { getCurrentInstance, type Ref, watch } from 'vue';

export class VueRay extends Ray {
    public component: any = { $data: {}, $refs: {} };
    public $watch: CallableFunction = () => {};

    public data(): void {
        if (this.component) {
            let data = Object.assign({}, this.component.data);

            // remove Vue Observers to avoid recursive issues
            delete data['trackingRays'];
            delete data['trackingStops'];

            data = Object.freeze(data);

            this.table([data], 'Data');
        }
    }

    public props(): void {
        if (this.component) {
            this.table([this.component.props]);
        }
    }

    public element(refName: string): VueRay {
        if (this.component?.refs[refName]) {
            return this.html((this.component?.refs[refName] as any).outerHTML);
        }

        return this;
    }

    public track(name: string): void {
        if (this.component && typeof this.component.trackingRays[name] === 'undefined') {
            this.component.trackingRays[name] = RayMixin.methods.$ray(this.component)(this.component.data['name']);

            const onTrackedUpdate = (value: any, previousValue: any) => {
                this.component.trackingRays[name] = this.component.trackingRays[name].sendCustom(
                    [
                        `<code style="font-weight: bold;">tracking: </code><code style="color: #16a34a;">${determineComponentNameDuringEvent(this.component.type)}.data.${name}</code>`,
                        `<hr width="33%" align="left" /><code style="color: #94a3b8;">new value: </code><code class="text-gray-500 bold font-bold">${prettyPrint(value)}</div>`,
                        `                               <code style="color: #94a3b8;">old value: </code><code class="text-gray-700">${prettyPrint(previousValue)}</code>`,
                        `<hr width="33%" align="left" /><code style="color: #94a3b8;">changed at: </code><code style="color: #60a5fa;">${new Date().toISOString()}</div>`,
                    ].join('<br>'),
                    'Tracking Data',
                );
            };

            this.component.trackingStops[name] = this.$watch(() => this.component.data[name], onTrackedUpdate, { deep: true });
        }
    }

    public untrack(name: string): void {
        if (this.component && typeof this.component.trackingStops[name] !== 'undefined') {
            this.component.trackingStops[name]();
            delete this.component.trackingStops[name];
        }

        if (this.component && typeof this.component.trackingRays[name] !== 'undefined') {
            delete this.component.trackingRays[name];
        }
    }

    public unwatch(name: string): void {
        if (this.component && typeof this.component.trackingStops[name] !== 'undefined') {
            this.component.trackingStops[name]();
            delete this.component.trackingStops[name];
        }

        if (this.component && typeof this.component.trackingRays[name] !== 'undefined') {
            delete this.component.trackingRays[name];
        }
    }

    public watch(name: string, ref: Ref<any>): void {
        if (this.component && typeof this.component.trackingRays[name] === 'undefined') {
            this.component.trackingRays[name] = RayMixin.methods.$ray(this.component)(ref.value);

            const onTrackedUpdate = (value: any, previousValue: any) => {
                this.component.trackingRays[name] = this.component.trackingRays[name].sendCustom(
                    [
                        `<code style="font-weight: bold;">watching: </code><code style="color: #16a34a;">${determineComponentNameDuringEvent(this.component.type)} âžž Ref&lt;${name}&gt;</code>`,
                        `<hr width="33%" align="left" /><code style="color: #94a3b8;">new value: </code><code class="text-gray-500 bold font-bold">${prettyPrint(value)}</div>`,
                        `                               <code style="color: #94a3b8;">old value: </code><code class="text-gray-700">${prettyPrint(previousValue)}</code>`,
                        `<hr width="33%" align="left" /><code style="color: #94a3b8;">changed at: </code><code style="color: #60a5fa" class="text-gray-500 bold font-bold">${new Date().toISOString()}</div>`,
                    ].join('<br>'),
                    'Watching Ref',
                );
            };

            this.component.trackingStops[name] = this.$watch(ref, onTrackedUpdate, { deep: true });
        }
    }

    public static useDefaultSettings(settings: any) {
        return super.useDefaultSettings(settings);
    }

    public send(...args: any[]) {
        return super.send(...args);
    }
}

export const ray = (...args: any[]) => {
    const result = VueRay.create() as VueRay;
    result.component = getCurrentInstance();
    result.$watch = watch;

    if (!args.length) {
        return result;
    }

    result.send(...args);

    return result;
};

/**
 * Returns a ray() function that is wrapped with the component instance
 *
 * @param component
 * @returns {(...args: any[]) => VueRay}
 */
// eslint-disable-next-line no-unused-vars
export function rayWrapped(component: any): (...args: any[]) => VueRay {
    // biome-ignore lint/complexity/useArrowFunction: <explanation>
    return function (...args: any[]) {
        const result = VueRay.create() as VueRay;

        result.component = component;
        result.$watch = watch;

        if (!args.length) {
            return result;
        }

        result.send(...args);

        return result;
    };
}
