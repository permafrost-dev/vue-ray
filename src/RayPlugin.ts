import { initializeOptions } from '@/InitializeOptions';
import { createPackageMetaProperty } from '@/lib/helpers';
import { rayWrapped } from '@/VueRay';
import type { App } from 'vue';

export const RayPlugin = {
    install: (app: App, options: any) => {
        createPackageMetaProperty(app.config.globalProperties);
        initializeOptions(options, app.config);

        app.provide('ray', options);
        app['$ray'] = rayWrapped(app);
    },
};
