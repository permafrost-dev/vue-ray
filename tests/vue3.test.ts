/* eslint-disable no-unused-vars */
import { VueRay } from '@/VueRay';
import { beforeEach, describe, expect, it } from 'vitest';
import { FakeRay } from './TestClasses/FakeRay';
import { FakeVm } from './TestClasses/FakeVm';
import { RayMixin } from '@/RayMixin';
import { RayPlugin } from '@/RayPlugin';

let ray: FakeRay, fakeApp: FakeApp;

class FakeApp {
    public providedItems: any[] = [];
    public config: any;
    public mixins: any = {};

    constructor() {
        this.providedItems = [];
        this.config = {
            globalProperties: {
                $ray: null,
            },
            errorHandler: null,
        };

        this.mixins = {};
        this.mixins.methods = {
            $ray: null,
        };
    }

    provide(name: any) {
        this.providedItems.push(name);
    }

    getProvided() {
        return this.providedItems;
    }

    getLastProvided() {
        return this.getProvided().slice(0)
            .pop();
    }

    mixin(obj: any) {
        this.mixins = Object.assign({}, this.mixins, obj);
    }
}

beforeEach(() => {
    ray = new FakeRay();
    fakeApp = new FakeApp();
});

const prepareMixin = () => {
    // @ts-ignore
    RayMixin.$options = {
        name: 'test',
        __file: 'test.js',
    };

    // @ts-ignore
    RayMixin.$ray = function (...args: any[]) {
        return ray.send(...args);
    };

    RayMixin.methods = {
        // @ts-ignore
        $ray(...args: any[]) {
            return ray.send(...args);
        },
    };
};

describe('Vue 3 Ray Plugin:', () => {
    it('installs successfully', () => {
        RayPlugin.install(fakeApp, {});

        expect(fakeApp.getLastProvided()).toBe('ray');
        expect(fakeApp.getProvided().length).toBe(1);
        //expect(fakeApp.config.globalProperties['$ray']).not.toBe(null);
        expect(fakeApp.mixins.methods.$ray).not.toBe(null);

        if (fakeApp.mixins.methods.$ray !== null) {
            const testRay = fakeApp.mixins.methods.$ray;

            expect(typeof testRay).not.toBe('undefined');
            expect(testRay().constructor.name).toBe('VueRay');
        }
    });

    it('installs successfully with the interceptErrors option', () => {
        RayPlugin.install(fakeApp, { interceptErrors: true });

        expect(fakeApp.getLastProvided()).toBe('ray');
        expect(fakeApp.getProvided().length).toBe(1);
        expect(fakeApp.mixins.methods.$ray).not.toBe(null);
        expect(fakeApp.config.errorHandler).not.toBeNull();
        expect(typeof fakeApp.config.errorHandler).toBe('function');

        const fakeVm = new FakeVm(ray);
        fakeApp.config.errorHandler({ stack: 'test error' }, fakeVm);

        expect(ray.payloads.length).toBe(3);
        expect(ray.payloads[0].content).toContain('test error');
        expect(ray.payloads[1].content).toBe('small');
        expect(ray.payloads[2].content).toBe('red');
    });
});

describe('Vue 3 Mixin:', () => {
    it('does not conditionally display events if not defined', () => {
        VueRay.show_component_lifecycles = [];

        prepareMixin();

        //RayMixin.beforeCreate();
        RayMixin.beforeMount();
        RayMixin.created();
        RayMixin.mounted();
        RayMixin.unmounted();
        RayMixin.updated();

        expect(ray.payloads.length).toBe(0);
    });

    it('conditionally displays the beforeMount event', () => {
        VueRay.show_component_lifecycles = ['before-mount'];

        prepareMixin();
        RayMixin.beforeMount();

        expect(ray.payloads.length).toBe(1);
        expect(ray.payloads).toMatchSnapshot();
    });

    it('conditionally displays the created & mounted events', () => {
        VueRay.show_component_lifecycles = ['created', 'mounted'];

        prepareMixin();
        RayMixin.created();
        RayMixin.mounted();

        expect(ray.payloads.length).toBe(2);
        expect(ray.payloads).toMatchSnapshot();
    });

    it('conditionally displays the unmounted event', () => {
        VueRay.show_component_lifecycles = ['unmounted'];

        prepareMixin();
        RayMixin.unmounted();

        expect(ray.payloads.length).toBe(1);
        expect(ray.payloads).toMatchSnapshot();
    });

    it('conditionally displays the updated event', () => {
        VueRay.show_component_lifecycles = ['updated'];

        prepareMixin();
        RayMixin.updated();

        expect(ray.payloads.length).toBe(1);
        expect(ray.payloads).toMatchSnapshot();
    });
});
