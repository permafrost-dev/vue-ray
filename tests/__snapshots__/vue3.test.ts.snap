// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Vue 3 Mixin: > conditionally displays the beforeMount event 1`] = `
[
  {
    "content": [
      "component before-mount: <code>test</code>",
      "filename: <code>&lt;project root&gt;/test.js</code>",
    ],
    "label": undefined,
    "name": "table",
  },
]
`;

exports[`Vue 3 Mixin: > conditionally displays the beforeUnmount event 1`] = `
[
  {
    "content": [
      "component before-unmount: <code>test</code>",
      "filename: <code>&lt;project root&gt;/test.js</code>",
    ],
    "label": undefined,
    "name": "table",
  },
]
`;

exports[`Vue 3 Mixin: > conditionally displays the created & mounted events 1`] = `
[
  {
    "content": [
      "component created: <code>test</code>",
      "filename: <code>&lt;project root&gt;/test.js</code>",
    ],
    "label": undefined,
    "name": "table",
  },
  {
    "content": [
      "component mounted: <code>test</code>",
      "filename: <code>&lt;project root&gt;/test.js</code>",
    ],
    "label": undefined,
    "name": "table",
  },
]
`;

exports[`Vue 3 Mixin: > conditionally displays the unmounted event 1`] = `
[
  {
    "content": [
      "component unmounted: <code>test</code>",
      "filename: <code>&lt;project root&gt;/test.js</code>",
    ],
    "label": undefined,
    "name": "table",
  },
]
`;

exports[`Vue 3 Mixin: > conditionally displays the updated event 1`] = `
[
  {
    "content": [
      "component updated: <code>test</code>",
      "filename: <code>&lt;project root&gt;/test.js</code>",
    ],
    "label": undefined,
    "name": "table",
  },
]
`;
